xl: Add '.CORRUPTED' prefix to corrupted objects

vadmeste opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
getObjectInfo() checks if a given object has enough xl.json in disks, returns 404 if not and added .CORRUPTED suffix to object.


## Motivation and Context
Fixes #6575 and #6591 

## Regression
No

## How Has This Been Tested?


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana reviewed (approved):
LGTM as agreed and tested.

kannappanr reviewed (dismissed):
LGTM

harshavardhana reviewed (changes requested):
I discussed with @abperiasamy the changes here are fine but we also have to make few more changes, i.e we should rename the object which are stale with "objectName.CORRUPTED" atomically. 

Also allow them to be available in the listing output such that they can be deleted by `mc rm`. 

What we need to do should be carefully understood here 

- First we read the `xl.json` from all disks and then ascertain how many disks its missing
- Then pick one of the common `xl.json` and decide the quorum. 
- Once the quorum is obtained see how many places the object still exists, rename all of them as objectName.corrupted
- Recreate empty objects on non-existent disks such as objectName.CORRUPTED with `xl.json` 
- This would allow for quorum and for the objects to be deleted.

All of this happens inside a ReadLock which is okay since other write locks will be blocked, we don't need to acquire a new write lock. Similar to healing but we are actively doing this during GetObjectInfo()

harshavardhana reviewed (changes requested):
is this PR ready?

vadmeste committed:
xl: Add '.CORRUPTED' prefix to corrupted objects

getObjectInfo() checks if a given object has enough xl.json in disks,
returns 404 if not and added .CORRUPTED suffix to object.

harshavardhana commented on cmd/xl-v1-object.go:
```diff
@@ -452,14 +453,35 @@ func (xl xlObjects) getObjectInfo(ctx context.Context, bucket, object string) (o
+			}
+		}
+		// If xl.json is not present in read quorum disks,
+		// add .CORRUPTION prefix to the current object.
```
.CORRUPTED typo here.

vadmeste replied:
> .CORRUPTED typo here.

Fixed

harshavardhana commented on cmd/xl-v1-object.go:
```diff
@@ -452,14 +453,35 @@ func (xl xlObjects) getObjectInfo(ctx context.Context, bucket, object string) (o
+	} else {
+		// If this is a corrupted object, change read quorum to N/2 disks
+		// so it will be visible to users, so they can delete it.
+		readQuorum = len(xl.getDisks()) / 2
```
We can keep the read quorum similar to standard? And not n/2 ?

vadmeste replied:
> We can keep the read quorum similar to standard? And not n/2 ?

We can't actually because otherwise the object won't be visible, the object is already corrupted anyway

harshavardhana commented on cmd/xl-v1-object.go:
```diff
@@ -452,14 +453,35 @@ func (xl xlObjects) getObjectInfo(ctx context.Context, bucket, object string) (o
+		// If xl.json is not present in read quorum disks,
+		// add .CORRUPTION prefix to the current object.
+		if len(xl.getDisks())-notFoundXLJSON < readQuorum {
+			rename(ctx, xl.getDisks(), bucket, object, bucket, object+xlCorruptedSuffix, true, len(xl.getDisks())/2+1, []error{errFileNotFound})
```
We can keep the quorum similar to global standard class ?

vadmeste replied:
> We can keep the quorum similar to global standard class ?

Fixed

harshavardhana reviewed (changes requested):
We have to be extra careful yes, but it is okay to make the object a user interactive responsibility.

We are not actively deleting anything so it is okay.

kannappanr reviewed (approved):
LGTM
